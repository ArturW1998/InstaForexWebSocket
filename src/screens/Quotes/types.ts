import {ViewStyle} from "react-native";

export type Style = {
    safeArea: ViewStyle,
};
