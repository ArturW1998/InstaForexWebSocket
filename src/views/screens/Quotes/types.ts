import { ViewStyle } from "react-native";

export type IStyle = {
    safeArea: ViewStyle,
};
