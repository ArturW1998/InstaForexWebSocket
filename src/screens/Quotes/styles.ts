import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  safeArea: {
    flex: 1,
    marginTop: 24,
  },
  container: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: 12,
    paddingVertical: 18,
    borderBottomWidth: 1,
    borderBottomColor: 'gray',
  },
  textContainer: {
    flex: 1,
    marginRight: 6,
  },
  changeFieldContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
  },
  symbol: {
    fontSize: 14,
    fontWeight: 'bold',
  },
  infoText: {
    fontSize: 14,
    fontWeight: '500',
    textAlign: 'right',
  },
  icon: {
    marginLeft: 4,
  },
});
