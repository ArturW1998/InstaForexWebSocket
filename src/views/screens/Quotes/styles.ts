import {StyleSheet} from 'react-native';
import { IStyle } from './types';

const styles = StyleSheet.create<IStyle>({
  safeArea: {
    flex: 1,
    marginTop: 24,
  },
});

export default styles;
