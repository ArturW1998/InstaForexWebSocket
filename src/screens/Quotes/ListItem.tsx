import React, {useEffect, useState} from 'react';
import {View, Text, Animated} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import {IQuota} from '../../types';
import styles from './styles';

type Props = {
  quota: IQuota;
};

export const ListItem = React.memo(({quota}: Props) => {
  const [animatedValue, setAnimatedValue] = useState(new Animated.Value(0));

  useEffect(() => {
    Animated.timing(animatedValue, {
      toValue: 1,
      duration: 1000,
      useNativeDriver: false,
    }).start(() => {
      setAnimatedValue(new Animated.Value(0));
    });
  }, [quota.change]);

  const interpolateColor = animatedValue.interpolate({
    inputRange: [0, 0.5, 1],
    outputRange: [
      'transparent',
      quota.change > 0 ? '#99ff99' : '#ff6666',
      'transparent',
    ],
  });

  const animatedStyle = {
    backgroundColor: interpolateColor,
  };
  const isNegative = quota.change < 0;

  return (
    <Animated.View style={[styles.container, animatedStyle]}>
      <View style={styles.leftContent}>
        <Text style={styles.symbol}>{quota.symbol}</Text>
      </View>
      <View style={styles.rightContent}>
        <View style={styles.textContainer}>
          <Text style={styles.infoText}>{quota.ask}</Text>
        </View>
        <View style={styles.textContainer}>
          <Text style={styles.infoText}>{quota.bid}</Text>
        </View>
        <View style={styles.changeFieldContainer}>
          <Text
            style={[
              styles.infoText,
              // eslint-disable-next-line react-native/no-inline-styles
              {color: isNegative ? 'red' : 'green'},
            ]}>
            {`${quota.change}%`}
          </Text>
          <Icon
            color={isNegative ? 'red' : 'green'}
            name={`chevron-${isNegative ? 'down' : 'up'}`}
            style={styles.icon}
          />
        </View>
      </View>
    </Animated.View>
  );
});
