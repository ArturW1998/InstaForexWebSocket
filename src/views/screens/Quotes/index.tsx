import React, { useContext, FC } from 'react';
import {SafeAreaView} from 'react-native';
import Themes from "../../../constants/themes";
import { QuotesContext } from '../../../contextes/quotesContext';
import QuotesList from '../../components/QuotesList';
import ErrorIndicator from "../../components/ErrorIndicator";
import AppLoader from '../../components/AppLoader';
import AppButton from '../../components/AppButton';
import { IContextValue } from '../../../types';
import styles from './styles';

const Quotes: FC = () => {
    const {
        data,
        isError,
        isLoading,
        fetchData,
    } = useContext<IContextValue>(QuotesContext);

    if (isError) {
        return (
            <ErrorIndicator>
                <AppButton backgroundColor={Themes.dangerColor} onPress={fetchData}>
                    Try again
                </AppButton>
            </ErrorIndicator>
        );
    }

    if (isLoading) return <AppLoader />;

  return (
    <SafeAreaView style={styles.safeArea}>
      <QuotesList quotes={data} />
    </SafeAreaView>
  );
};

export default Quotes;
