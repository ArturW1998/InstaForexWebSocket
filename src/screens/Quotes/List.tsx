import React, {useCallback} from 'react';
import {FlatList} from 'react-native';

import {IQuota} from '../../types';
import {ListItem} from './ListItem';

type Props = {
  quotes: IQuota[];
};

export const QuotesList = ({quotes}: Props) => {
  const renderItem = useCallback(({item}: {item: IQuota}) => {
    return <ListItem quota={item} />;
  }, []);

  const keyExtractor = useCallback((item) => item.symbol, []);

  return (
    <FlatList
      data={quotes}
      renderItem={renderItem}
      keyExtractor={keyExtractor}
    />
  );
};
