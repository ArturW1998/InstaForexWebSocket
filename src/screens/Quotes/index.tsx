import React, {FC} from 'react';
import {SafeAreaView} from 'react-native';
import useFetchQuotes from "../../hooks/useFetchQuotes";
import {QuotesList} from './List';
import styles from './styles';

const Quotes: FC = () => {
    const [quotes, fetchData] = useFetchQuotes();

  return (
    <SafeAreaView style={styles.safeArea}>
      <QuotesList quotes={quotes} />
    </SafeAreaView>
  );
};

export default Quotes;
