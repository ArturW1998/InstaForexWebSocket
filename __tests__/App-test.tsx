import React from 'react';
import renderer from 'react-test-renderer';
import Quotes from '../src/views/screens/Quotes';

describe('App', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Quotes />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
