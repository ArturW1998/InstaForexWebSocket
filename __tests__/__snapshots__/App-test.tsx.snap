// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "marginTop": 24,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 300,
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": "100%",
              "resizeMode": "contain",
              "width": "100%",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
